cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp

```diff
@@ -1092,139 +1721,151 @@ void WindowBlockManager::offloadBlock(
 
-            numMatchedTokens += numMatched > 0 ? numMatched : blockItr->uniqueTokens.size();
+            numMatchedTokens += numMatched > 0 ? numMatched : matchingNode->getUniqueTokens().size();
             if (perBlockRetentions[bi].retentionPriority.has_value()
                 && matchingBlock->getPriority() != perBlockRetentions[bi].retentionPriority && mEventManager)
             {
+                TLLM_LOG_DEBUG("%s;%d - mEventManager->enqueueUpdatedEvent(block=%d) Priority changed from %d to %d",
+                    __FILE__, __LINE__, matchingBlock->getBlockId(), matchingBlock->getPriority(),
+                    *perBlockRetentions[bi].retentionPriority);
                 mEventManager->enqueueUpdatedEvent(
                     tle::KVCacheUpdatedData(matchingBlock->getHash())
                         .priorityUpdated(matchingBlock->getPriority(), *perBlockRetentions[bi].retentionPriority),
                     mWindowSize);
             }
             if (partialMatch)
             {
-                if (matchingBlock->hasRefs() || !matchingBlock->isLeaf())
+                auto partialBlockKey = matchingNode->getBlockKey().clone(numMatched);
+                if (matchingBlock->hasRefs() || !matchingNode->isLeaf())
                 {
-                    // Somebody else is using block or it is not a leaf, copy reusable tokens
-                    auto newBlock = getFreeBlock(
-                        sequence, matchingBlock->getPriority(), matchingBlock->getDurationMs(), mode, directory);
+                    // Somebody else is using block or this is full attention and block is not a leaf, copy reusable
+                    // tokens
+                    // TODO: Consider whether non-leaf blocks should be reuse instead of copied for SWA layers.
+                    auto newBlock
+                        = getFreeBlock(matchingBlock->getPriority(), matchingBlock->getDurationMs(), mode, directory);
                     mTransferManager->onboard(matchingBlock, newBlock, mPools, numMatched, mode, directory);
                     // TODO: (optional) Send out event
-                    matchingBlock = newBlock;
-                    if (blockItr != blockKeys.end())
+                    TLLM_LOG_DEBUG(
+                        "%s::loadOrAllocateBlocks - Copied partially filled block %d into %d. Reserved block %d for "
+                        "sequence %lu",
+                        mLogPrefix.c_str(), matchingBlock->getBlockId(), newBlock->getBlockId(), newBlock->getBlockId(),
+                        sequence.getRequestId());
+                    if (!matchingBlock->hasRefs())
                     {
-                        matchingBlock->setBlockKey(
-                            *blockItr, blockItr->uniqueTokens.size() == static_cast<size_t>(mTokensPerBlock));
+                        // Release the matching block
+                        // TODO: Pending copy of matchingBlock to newBlock. Can releasing it cause race condition?
```

Comment:
I want to understand more upon the technical details here.

If I understand correctly, the kv cache manager operates on a single thread, and we iterate upon the context/generation requests for kv cache block manipulation. Is the race condition mentioned here specific to dis-aggregated serving where we have two threads there? Or what other conditions may cause a race condition?